<a href="https://colab.research.google.com/github/Saurav-1399/Coursera_Capstone/blob/main/The_Battle_of_Neighborhoods_2-Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#                                                                   The Battle of Neighborhoods | Segmentation & Clustering of Neighborhoods in London & Paris

Author: Saurav Menon M

Date: 13th November 2020

# Introduction

Travel has been a part of our lives for a long time. Travel Experiences give our travelers the ability to experience the local culture while traveling in a relaxed way. Traveling is one of the best way to enhance personal growth. It enables you to do things different from your daily routine activities. When you travel, you step out from your comfort zone to a different environment which makes you become more responsible and gives a sense of independence.

London and Paris are quite the popular tourist and vacation destinations for people all around the world. They are diverse and multicultural and offer a wide variety of experiences that is widely sought after. We try to group the neighbourhoods of London and Paris respectively and draw insights to what they look like now.

# Business Problem

The aim is to help tourists choose their destinations depending on the experiences that the neighbourhoods have to offer and what they would want to have. This also helps people make decisions if they are thinking about migrating to London or Paris or even if they want to relocate neighbourhoods within the city. Our findings will help stakeholders make informed decisions and address any concerns they have including the different kinds of cuisines, provision stores and what the city has to offer.

# Data Description

We require geolocation data for both London and Paris. Postal codes in each city serve as a starting point. Using Postal codes we use can find out the neighbourhoods, boroughs, venues and their most popular venue categories.

## London
To derive our solution, We scrape our data from https://en.wikipedia.org/wiki/List_of_areas_of_London

This wikipedia page has information about all the neighbourhoods, we limit it London.

1. borough : Name of Neighbourhood
2. town : Name of borough
3. post_code : Postal codes for London.

This wikipedia page lacks information about the geographical locations. To solve this problem we use ArcGIS API

## ArcGIS API

ArcGIS Online enables you to connect people, locations, and data using interactive maps. Work with smart, data-driven styles and intuitive analysis tools that deliver location intelligence. Share your insights with the world or specific groups.

More specifically, we use ArcGIS to get the geo locations of the neighbourhoods of London. The following columns are added to our initial dataset which prepares our data.

1. latitude : Latitude for Neighbourhood
2. longitude : Longitude for Neighbourhood

## Paris

To derive our solution, We leverage JSON data available at https://www.data.gouv.fr/fr/datasets/r/e88c6fda-1d09-42a0-a069-606d3259114e

The JSON file has data about all the neighbourhoods in France, we limit it to Paris.

1. postal_code : Postal codes for France
2. nom_comm : Name of Neighbourhoods in France
3. nom_dept : Name of the boroughs, equivalent to towns in France
4. geo_point_2d : Tuple containing the latitude and longitude of the Neighbourhoods.

## Foursquare API Data

We will need data about different venues in different neighbourhoods of that specific borough. In order to gain that information we will use "Foursquare" locational information. Foursquare is a location data provider with information about all manner of venues and events within an area of interest. Such information includes venue names, locations, menus and even photos. As such, the foursquare location platform will be used as the sole data source since all the stated required information can be obtained through the API.

After finding the list of neighbourhoods, we then connect to the Foursquare API to gather information about venues inside each and every neighbourhood. For each neighbourhood, we have chosen the radius to be 500 meters.

The data retrieved from Foursquare contained information of venues within a specified distance of the longitude and latitude of the postcodes. The information obtained per venue as follows:

1. Neighbourhood : Name of the Neighbourhood
2. Neighbourhood Latitude : Latitude of the Neighbourhood
3. Neighbourhood Longitude : Longitude of the Neighbourhood
4. Venue : Name of the Venue
5. Venue Latitude : Latitude of Venue
6. Venue Longitude : Longitude of Venue
7. Venue Category : Category of Venue

Based on all the information collected for both London and Paris, we have sufficient data to build our model. We cluster the neighbourhoods together based on similar venue categories. We then present our observations and findings. Using this data, our stakeholders can take the necessary decision.

# Methodology

We will be creating our model with the help of Python so we start off by importing all the required packages.

In [ ]:
!pip install folium

     |████████████████████████████████| 93 kB 3.6 MB/s eta 0:00:011


In [ ]:
import pandas as pd
import requests
import numpy as np
import matplotlib.cm as cm
import matplotlib.colors as colors
import folium

# import k-means for the clustering stage
from sklearn.cluster import KMeans

The approach taken here is to explore each of the cities individually, plot the map to show the neighbourhoods being considered and then build our model by clustering all of the similar neighbourhoods together and finally plot the new map with the clustered neighbourhoods. We draw insights and then compare and discuss our findings.

# Exploring London

### Neighbourhoods of London

We begin to start collecting and refining the data needed for the our business solution to work.

### Data Collection

To get the neighbourhoods in london, we start by scraping the list of areas of london wiki page.

In [ ]:
url_london = "https://en.wikipedia.org/wiki/List_of_areas_of_London"
wiki_london_url = requests.get(url_london)
wiki_london_url

<Response [200]>

Response 200 means that we are able to make the connection

In [ ]:
wiki_london_data = pd.read_html(wiki_london_url.text)
wiki_london_data

[                                                   0
 0  Map all coordinates in "Category:Areas of Lond...
 1                 Download coordinates as: KML · GPX,
             Location                     London borough       Post town  \
 0         Abbey Wood              Bexley, Greenwich [7]          LONDON   
 1              Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
 2          Addington                         Croydon[8]         CROYDON   
 3         Addiscombe                         Croydon[8]         CROYDON   
 4        Albany Park                             Bexley  BEXLEY, SIDCUP   
 ..               ...                                ...             ...   
 527         Woolwich                          Greenwich          LONDON   
 528   Worcester Park       Sutton, Kingston upon Thames  WORCESTER PARK   
 529  Wormwood Scrubs             Hammersmith and Fulham          LONDON   
 530          Yeading                         Hillingdon           HAYES   
 

Scraping the webpage gives us all the tables present on the page. We need the 2nd table, so selecting the 2nd table.

In [ ]:
wiki_london_data = wiki_london_data[1]
wiki_london_data

Location                     London borough       Post town  \
0         Abbey Wood              Bexley, Greenwich [7]          LONDON   
1              Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
2          Addington                         Croydon[8]         CROYDON   
3         Addiscombe                         Croydon[8]         CROYDON   
4        Albany Park                             Bexley  BEXLEY, SIDCUP   
..               ...                                ...             ...   
527         Woolwich                          Greenwich          LONDON   
528   Worcester Park       Sutton, Kingston upon Thames  WORCESTER PARK   
529  Wormwood Scrubs             Hammersmith and Fulham          LONDON   
530          Yeading                         Hillingdon           HAYES   
531         Yiewsley                         Hillingdon    WEST DRAYTON   

    Postcode district Dial code OS grid ref  
0                 SE2       020    TQ465785  
1              W3, W4       020    TQ205805  
2                 CR0       020    TQ375645  
3                 CR0       020    TQ345665  
4           DA5, DA14       020    TQ478728  
..                ...       ...         ...  
527              SE18       020    TQ435795  
528               KT4       020    TQ225655  
529               W12       020    TQ225815  
530               UB4       020    TQ115825  
531               UB7       020    TQ063804  

[532 rows x 6 columns]

### Data Preprocessing

we remove the spaces in the column titles and then we add _ between words.

In [ ]:
wiki_london_data.rename(columns=lambda x: x.strip().replace(" ", "_"), inplace=True)
wiki_london_data

Location                     London borough       Post_town  \
0         Abbey Wood              Bexley, Greenwich [7]          LONDON   
1              Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
2          Addington                         Croydon[8]         CROYDON   
3         Addiscombe                         Croydon[8]         CROYDON   
4        Albany Park                             Bexley  BEXLEY, SIDCUP   
..               ...                                ...             ...   
527         Woolwich                          Greenwich          LONDON   
528   Worcester Park       Sutton, Kingston upon Thames  WORCESTER PARK   
529  Wormwood Scrubs             Hammersmith and Fulham          LONDON   
530          Yeading                         Hillingdon           HAYES   
531         Yiewsley                         Hillingdon    WEST DRAYTON   

    Postcode district Dial code OS_grid_ref  
0                 SE2       020    TQ465785  
1              W3, W4       020    TQ205805  
2                 CR0       020    TQ375645  
3                 CR0       020    TQ345665  
4           DA5, DA14       020    TQ478728  
..                ...       ...         ...  
527              SE18       020    TQ435795  
528               KT4       020    TQ225655  
529               W12       020    TQ225815  
530               UB4       020    TQ115825  
531               UB7       020    TQ063804  

[532 rows x 6 columns]

We see that few columns have no '_' between the words despite applying our function meaning that there are special characters

### Feature Selection

We need only the boroughs, Postal codes, Post town for further steps. We can drop the locations, dial codes and OS grid.

In [ ]:
df1 = wiki_london_data.drop( [ wiki_london_data.columns[0], wiki_london_data.columns[4], wiki_london_data.columns[5] ], axis=1)

In [ ]:
df1.head()

London borough       Post_town Postcode district
0              Bexley, Greenwich [7]          LONDON               SE2
1  Ealing, Hammersmith and Fulham[8]          LONDON            W3, W4
2                         Croydon[8]         CROYDON               CR0
3                         Croydon[8]         CROYDON               CR0
4                             Bexley  BEXLEY, SIDCUP         DA5, DA14

let's rename the Postcode district column and the london borough to something simpler

In [ ]:
df1.columns = ['borough','town','post_code']
df1

borough            town  post_code
0                Bexley, Greenwich [7]          LONDON        SE2
1    Ealing, Hammersmith and Fulham[8]          LONDON     W3, W4
2                           Croydon[8]         CROYDON        CR0
3                           Croydon[8]         CROYDON        CR0
4                               Bexley  BEXLEY, SIDCUP  DA5, DA14
..                                 ...             ...        ...
527                          Greenwich          LONDON       SE18
528       Sutton, Kingston upon Thames  WORCESTER PARK        KT4
529             Hammersmith and Fulham          LONDON        W12
530                         Hillingdon           HAYES        UB4
531                         Hillingdon    WEST DRAYTON        UB7

[532 rows x 3 columns]

Let's remove the Square brackets [ ] and numbers from the borough column

In [ ]:
df1['borough'] = df1['borough'].map(lambda x: x.rstrip(']').rstrip('0123456789').rstrip('['))
df1

borough            town  post_code
0                Bexley, Greenwich           LONDON        SE2
1    Ealing, Hammersmith and Fulham          LONDON     W3, W4
2                           Croydon         CROYDON        CR0
3                           Croydon         CROYDON        CR0
4                            Bexley  BEXLEY, SIDCUP  DA5, DA14
..                              ...             ...        ...
527                       Greenwich          LONDON       SE18
528    Sutton, Kingston upon Thames  WORCESTER PARK        KT4
529          Hammersmith and Fulham          LONDON        W12
530                      Hillingdon           HAYES        UB4
531                      Hillingdon    WEST DRAYTON        UB7

[532 rows x 3 columns]

Take the dimension of the dataframe

In [ ]:
df1.shape

(532, 3)

We currently have 532 records and 3 columns of our data. It's time to perform Feature Engineering

### Feature Engineering

We can only focusing on the neighbourhoods of London, so performing the changes

In [ ]:
df1 = df1[df1['town'].str.contains('LONDON')]
df1

borough                    town post_code
0                Bexley, Greenwich                   LONDON       SE2
1    Ealing, Hammersmith and Fulham                  LONDON    W3, W4
6                              City                  LONDON       EC3
7                       Westminster                  LONDON       WC2
9                           Bromley                  LONDON      SE20
..                              ...                     ...       ...
522                       Redbridge                  LONDON  IG8, E18
523       Redbridge, Waltham Forest  LONDON, WOODFORD GREEN       IG8
526                          Barnet                  LONDON       N12
527                       Greenwich                  LONDON      SE18
529          Hammersmith and Fulham                  LONDON       W12

[309 rows x 3 columns]

In [ ]:
df1.shape

(309, 3)

We now have only 309 rows. We can proceed with our further steps. Getting some descriptive statistics

In [ ]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 309 entries, 0 to 529
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   borough    309 non-null    object
 1   town       309 non-null    object
 2   post_code  309 non-null    object
dtypes: object(3)
memory usage: 9.7+ KB


## Geolocations of the London Neighbourhoods

### ArcGis API

We need to get the geographical co-ordinates for the neighbourhoods to plot out map. We will use the arcgis package to do so.

Arcgis doesn't have a limitation on the number of API calls made so it fits our use case perfectly.

In [ ]:
!pip install arcgis

In [ ]:
from arcgis.geocoding import geocode
from arcgis.gis import GIS
gis = GIS()

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/arcgis/features/_data/geodataset/utils.py:16: FutureWarning: The pandas.datetime class is deprecated and will be removed from pandas in a future version. Import from datetime module instead.
  pd.datetime,


Defining London arcgis geocode function to return latitude and longitude

In [ ]:
def get_x_y_uk(address1):
   lat_coords = 0
   lng_coords = 0
   g = geocode(address='{}, London, England, GBR'.format(address1))[0]
   lng_coords = g['location']['x']
   lat_coords = g['location']['y']
   return str(lat_coords) +","+ str(lng_coords)

Checking sample data

In [ ]:
c = get_x_y_uk('SE2')

In [ ]:
c

'51.492450000000076,0.12127000000003818'

Looks good, We Copy over the postal codes of london to pass it into the geolocator function that we just defined above

In [ ]:
geo_coordinates_uk = df1['post_code']    
geo_coordinates_uk

0           SE2
1        W3, W4
6           EC3
7           WC2
9          SE20
         ...   
522    IG8, E18
523         IG8
526         N12
527        SE18
529         W12
Name: post_code, Length: 309, dtype: object

Passing postal codes of london to get the geographical co-ordinates

In [ ]:
coordinates_latlng_uk = geo_coordinates_uk.apply(lambda x: get_x_y_uk(x))
coordinates_latlng_uk

0       51.492450000000076,0.12127000000003818
1        51.51324000000005,-0.2674599999999714
6       51.51200000000006,-0.08057999999994081
7       51.51651000000004,-0.11967999999995982
9       51.41009000000008,-0.05682999999993399
                        ...                   
522    51.589770000000044,0.030520000000024083
523      51.50642000000005,-0.1272099999999341
526     51.615920000000074,-0.1767399999999384
527      51.48207000000008,0.07143000000002075
529      51.50645000000003,-0.2369099999999662
Name: post_code, Length: 309, dtype: object

### Latitude

Extracting the latitude from our previously collected coordinates

In [ ]:
lat_uk = coordinates_latlng_uk.apply(lambda x: x.split(',')[0])
lat_uk

0      51.492450000000076
1       51.51324000000005
6       51.51200000000006
7       51.51651000000004
9       51.41009000000008
              ...        
522    51.589770000000044
523     51.50642000000005
526    51.615920000000074
527     51.48207000000008
529     51.50645000000003
Name: post_code, Length: 309, dtype: object

### Longitude

Extracting the Longitude from our previously collected coordinates

In [ ]:
lng_uk = coordinates_latlng_uk.apply(lambda x: x.split(',')[1])
lng_uk

0       0.12127000000003818
1       -0.2674599999999714
6      -0.08057999999994081
7      -0.11967999999995982
9      -0.05682999999993399
               ...         
522    0.030520000000024083
523     -0.1272099999999341
526     -0.1767399999999384
527     0.07143000000002075
529     -0.2369099999999662
Name: post_code, Length: 309, dtype: object

We now have the geographical co-ordinates of the London Neighbourhoods.

We proceed with Merging our source data with the geographical co-ordinates to make our dataset ready for the next stage

In [ ]:
london_merged = pd.concat([df1,lat_uk.astype(float), lng_uk.astype(float)], axis=1)
london_merged.columns= ['borough','town','post_code','latitude','longitude']
london_merged

borough                    town post_code  \
0                Bexley, Greenwich                   LONDON       SE2   
1    Ealing, Hammersmith and Fulham                  LONDON    W3, W4   
6                              City                  LONDON       EC3   
7                       Westminster                  LONDON       WC2   
9                           Bromley                  LONDON      SE20   
..                              ...                     ...       ...   
522                       Redbridge                  LONDON  IG8, E18   
523       Redbridge, Waltham Forest  LONDON, WOODFORD GREEN       IG8   
526                          Barnet                  LONDON       N12   
527                       Greenwich                  LONDON      SE18   
529          Hammersmith and Fulham                  LONDON       W12   

     latitude  longitude  
0    51.49245    0.12127  
1    51.51324   -0.26746  
6    51.51200   -0.08058  
7    51.51651   -0.11968  
9    51.41009   -0.05683  
..        ...        ...  
522  51.58977    0.03052  
523  51.50642   -0.12721  
526  51.61592   -0.17674  
527  51.48207    0.07143  
529  51.50645   -0.23691  

[309 rows x 5 columns]

In [ ]:
london_merged.dtypes

borough       object
town          object
post_code     object
latitude     float64
longitude    float64
dtype: object

### Co-ordinates for London

Getting the geocode for London to help visualize it on the map

In [ ]:
london = geocode(address='London, England, GBR')[0]
london_lng_coords = london['location']['x']
london_lat_coords = london['location']['y']
london_lng_coords

-0.1272099999999341

In [ ]:
london_lat_coords

51.50642000000005

## Visualize the Map of London

To help visualize the Map of London and the neighbourhoods in London, we make use of the folium package.

In [ ]:
# Creating the map of London
map_London = folium.Map(location=[london_lat_coords, london_lng_coords], zoom_start=12)
map_London

# adding markers to map
for latitude, longitude, borough, town in zip(london_merged['latitude'], london_merged['longitude'], london_merged['borough'], london_merged['town']):
    label = '{}, {}'.format(town, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_London)  
    
map_London

### Venues in London

To proceed with the next part, we need to define Foursquare API credentials.

Using Foursquare API, we are able to get the venue and venue categories around each neighbourhood in London.

In [ ]:
CLIENT_ID = 'XDRQZ5F5BIHTJFQRHDVCWAFAGGZX3BTRIPVBRVLRP143YBFM' 
CLIENT_SECRET = 'XH1RB0NQ1YHMITFP1NUCER2EABPJAXYNCBX4IPE0IIPHUGQT'
VERSION = '20180605' # Foursquare API version

Defining a function to get the neraby venues in the neighbourhood. This will help us get venue categories which is important for our analysis

In [ ]:
LIMIT=100

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius,
            LIMIT
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Getting the venues in London

In [ ]:
venues_in_London = getNearbyVenues(london_merged['borough'], london_merged['latitude'], london_merged['longitude'])

Bexley, Greenwich 
Ealing, Hammersmith and Fulham
City
Westminster
Bromley
Islington
Islington
Barnet
Enfield
Wandsworth
Southwark
City
Richmond upon Thames
Barnet
Islington
Wandsworth
Westminster
Bromley
Newham
Ealing
Westminster
Lewisham
Camden
Southwark
Tower Hamlets
Bexley
City
Lewisham
Greenwich
Tower Hamlets
Camden
Haringey
Tower Hamlets
Haringey
Barnet
Brent
Lambeth
Lewisham
Tower Hamlets
Kensington and ChelseaHammersmith and Fulham
Brent
Barnet
Barnet
Southwark
Tower Hamlets
Camden
Tower Hamlets
Waltham Forest
Newham
Islington
Richmond upon Thames
Lewisham
Camden
Westminster
Greenwich
Kensington and Chelsea
Barnet
Westminster
Lewisham
Waltham Forest
Hounslow, Ealing, Hammersmith and Fulham
Brent
Barnet
Lambeth, Wandsworth
Islington
Barnet
Merton
Barnet
Westminster
Barnet, Brent, Camden
Lewisham
Bexley
Haringey
Bromley
Tower Hamlets
Newham
Hackney
Islington
Southwark
Lewisham
Brent
Southwark
Ealing
Kensington and Chelsea
Wandsworth
Southwark
Barnet
Newham
Richmond upon Thames
En

In [ ]:
venues_in_London.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0  Bexley, Greenwich                 51.49245                  0.12127   
1  Bexley, Greenwich                 51.49245                  0.12127   
2  Bexley, Greenwich                 51.49245                  0.12127   
3  Bexley, Greenwich                 51.49245                  0.12127   
4  Bexley, Greenwich                 51.49245                  0.12127   

                              Venue Venue Category  
0                      Lesnes Abbey  Historic Site  
1                       Sainsbury's    Supermarket  
2                              Lidl    Supermarket  
3  Abbey Wood Railway Station (ABW)  Train Station  
4                       Bean @ Work    Coffee Shop

In [ ]:
venues_in_London.shape

(10504, 5)

Wow, we have scraped together 10504 records for venues. This will definitely make the clustering interesting.

### Grouping by Venue Categories

We need to now see how many Venue Categories are there for further processing

In [ ]:
venues_in_London.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                      
Adult Boutique                                          Islington   
African Restaurant                                    Westminster   
American Restaurant                                Waltham Forest   
Antique Shop                                          Westminster   
Arcade                                                Westminster   
...                                                           ...   
Wings Joint                                Hammersmith and Fulham   
Women's Store        Kensington and ChelseaHammersmith and Fulham   
Xinjiang Restaurant                                     Southwark   
Yoga Studio                                           Westminster   
Zoo Exhibit                                                Camden   

                     Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                         
Adult Boutique                     51.52969                 -0.08697   
African Restaurant                 51.52587                 -0.08808   
American Restaurant                51.63261                  0.02795   
Antique Shop                       51.51651                 -0.11968   
Arcade                             51.51651                 -0.11968   
...                                     ...                      ...   
Wings Joint                        51.54187                 -0.19795   
Women's Store                      51.55457                 -0.11478   
Xinjiang Restaurant                51.47480                 -0.09313   
Yoga Studio                        51.55457                 -0.03558   
Zoo Exhibit                        51.53354                 -0.14606   

                                           Venue  
Venue Category                                    
Adult Boutique       Sh! Women's Erotic Emporium  
African Restaurant            Red Sea Restaurant  
American Restaurant                  Spielburger  
Antique Shop            The London Silver Vaults  
Arcade                        Novelty Automation  
...                                          ...  
Wings Joint                             Wingmans  
Women's Store                 Vivien of Holloway  
Xinjiang Restaurant                    Silk Road  
Yoga Studio                            yogahaven  
Zoo Exhibit                        Penguin Beach  

[293 rows x 4 columns]

We can see 293 records, just goes to show how diverse and interesting the place is.

### One Hot Encoding

We need to Encode our venue categories to get a better result for our clustering

In [ ]:
London_venue_cat = pd.get_dummies(venues_in_London[['Venue Category']], prefix="", prefix_sep="")
London_venue_cat

Adult Boutique  African Restaurant  American Restaurant  Antique Shop  \
0                   0                   0                    0             0   
1                   0                   0                    0             0   
2                   0                   0                    0             0   
3                   0                   0                    0             0   
4                   0                   0                    0             0   
...               ...                 ...                  ...           ...   
10499               0                   0                    0             0   
10500               0                   0                    0             0   
10501               0                   0                    0             0   
10502               0                   0                    0             0   
10503               0                   0                    0             0   

       Arcade  Arepa Restaurant  Argentinian Restaurant  Art Gallery  \
0           0                 0                       0            0   
1           0                 0                       0            0   
2           0                 0                       0            0   
3           0                 0                       0            0   
4           0                 0                       0            0   
...       ...               ...                     ...          ...   
10499       0                 0                       0            0   
10500       0                 0                       0            0   
10501       0                 0                       0            0   
10502       0                 0                       0            0   
10503       0                 0                       0            0   

       Art Museum  Arts & Crafts Store  ...  Vietnamese Restaurant  \
0               0                    0  ...                      0   
1               0                    0  ...                      0   
2               0                    0  ...                      0   
3               0                    0  ...                      0   
4               0                    0  ...                      0   
...           ...                  ...  ...                    ...   
10499           0                    0  ...                      0   
10500           0                    0  ...                      0   
10501           0                    0  ...                      0   
10502           0                    0  ...                      0   
10503           0                    0  ...                      0   

       Warehouse Store  Whisky Bar  Wine Bar  Wine Shop  Wings Joint  \
0                    0           0         0          0            0   
1                    0           0         0          0            0   
2                    0           0         0          0            0   
3                    0           0         0          0            0   
4                    0           0         0          0            0   
...                ...         ...       ...        ...          ...   
10499                0           0         0          0            0   
10500                0           0         0          0            0   
10501                0           0         0          0            0   
10502                0           0         0          0            0   
10503                0           0         0          0            0   

       Women's Store  Xinjiang Restaurant  Yoga Studio  Zoo Exhibit  
0                  0                    0            0            0  
1                  0                    0            0            0  
2                  0                    0            0            0  
3                  0                    0            0            0  
4                  0                    0            0            0  
...              ...                  ...          ... 

Adding Neighbourhood into the mix.

In [ ]:
London_venue_cat['Neighbourhood'] = venues_in_London['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [London_venue_cat.columns[-1]] + list(London_venue_cat.columns[:-1])
London_venue_cat = London_venue_cat[fixed_columns]

London_venue_cat.head()

Neighbourhood  Adult Boutique  African Restaurant  \
0  Bexley, Greenwich                0                   0   
1  Bexley, Greenwich                0                   0   
2  Bexley, Greenwich                0                   0   
3  Bexley, Greenwich                0                   0   
4  Bexley, Greenwich                0                   0   

   American Restaurant  Antique Shop  Arcade  Arepa Restaurant  \
0                    0             0       0                 0   
1                    0             0       0                 0   
2                    0             0       0                 0   
3                    0             0       0                 0   
4                    0             0       0                 0   

   Argentinian Restaurant  Art Gallery  Art Museum  ...  \
0                       0            0           0  ...   
1                       0            0           0  ...   
2                       0            0           0  ...   
3                       0            0           0  ...   
4                       0            0           0  ...   

   Vietnamese Restaurant  Warehouse Store  Whisky Bar  Wine Bar  Wine Shop  \
0                      0                0           0         0          0   
1                      0                0           0         0          0   
2                      0                0           0         0          0   
3                      0                0           0         0          0   
4                      0                0           0         0          0   

   Wings Joint  Women's Store  Xinjiang Restaurant  Yoga Studio  Zoo Exhibit  
0            0              0                    0            0            0  
1            0              0                    0            0            0  
2            0              0                    0            0            0  
3            0              0                    0            0            0  
4            0              0                    0            0            0  

[5 rows x 294 columns]

### Venue categories mean value

We will group the Neighbourhoods and calculate the mean venue categories value in each Neighbourhood

In [ ]:
London_grouped = London_venue_cat.groupby('Neighbourhood').mean().reset_index()
London_grouped.head()

Neighbourhood  Adult Boutique  African Restaurant  \
0                 Barnet             0.0                 0.0   
1  Barnet, Brent, Camden             0.0                 0.0   
2                 Bexley             0.0                 0.0   
3      Bexley, Greenwich             0.0                 0.0   
4     Bexley, Greenwich              0.0                 0.0   

   American Restaurant  Antique Shop  Arcade  Arepa Restaurant  \
0              0.00713           0.0     0.0               0.0   
1              0.00000           0.0     0.0               0.0   
2              0.00000           0.0     0.0               0.0   
3              0.00000           0.0     0.0               0.0   
4              0.00000           0.0     0.0               0.0   

   Argentinian Restaurant  Art Gallery  Art Museum  ...  \
0                 0.00713          0.0         0.0  ...   
1                 0.00000          0.0         0.0  ...   
2                 0.00000          0.0         0.0  ...   
3                 0.00000          0.0         0.0  ...   
4                 0.00000          0.0         0.0  ...   

   Vietnamese Restaurant  Warehouse Store  Whisky Bar  Wine Bar  Wine Shop  \
0                0.00713              0.0         0.0       0.0        0.0   
1                0.00000              0.0         0.0       0.0        0.0   
2                0.00000              0.0         0.0       0.0        0.0   
3                0.00000              0.0         0.0       0.0        0.0   
4                0.00000              0.0         0.0       0.0        0.0   

   Wings Joint  Women's Store  Xinjiang Restaurant  Yoga Studio  Zoo Exhibit  
0          0.0            0.0                  0.0          0.0          0.0  
1          0.0            0.0                  0.0          0.0          0.0  
2          0.0            0.0                  0.0          0.0          0.0  
3          0.0            0.0                  0.0          0.0          0.0  
4          0.0            0.0                  0.0          0.0          0.0  

[5 rows x 294 columns]

Let's make a function to get the top most common venue categories

In [ ]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods.

Creating a function to label the columns of the venue correctly

In [ ]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

### Top venue categories

Getting the top venue categories in London

In [ ]:
# create a new dataframe for London
neighborhoods_venues_sorted_london = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted_london['Neighbourhood'] = London_grouped['Neighbourhood']

for ind in np.arange(London_grouped.shape[0]):
    neighborhoods_venues_sorted_london.iloc[ind, 1:] = return_most_common_venues(London_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted_london.head()

Neighbourhood 1st Most Common Venue 2nd Most Common Venue  \
0                 Barnet           Coffee Shop                  Café   
1  Barnet, Brent, Camden           Bus Station  Gym / Fitness Center   
2                 Bexley           Supermarket         Historic Site   
3      Bexley, Greenwich           Sports Club     Convenience Store   
4     Bexley, Greenwich            Supermarket         Train Station   

  3rd Most Common Venue 4th Most Common Venue       5th Most Common Venue  \
0         Grocery Store    Italian Restaurant                         Pub   
1        Clothing Store           Supermarket                 Zoo Exhibit   
2           Coffee Shop     Convenience Store               Train Station   
3                  Park         Historic Site  Construction & Landscaping   
4     Convenience Store           Coffee Shop               Historic Site   

  6th Most Common Venue 7th Most Common Venue       8th Most Common Venue  \
0              Bus Stop           Supermarket            Sushi Restaurant   
1     Fish & Chips Shop    Falafel Restaurant              Farmers Market   
2                  Park              Bus Stop  Construction & Landscaping   
3           Golf Course              Bus Stop                 Flower Shop   
4     Fish & Chips Shop    Falafel Restaurant              Farmers Market   

  9th Most Common Venue 10th Most Common Venue  
0              Pharmacy     Turkish Restaurant  
1  Fast Food Restaurant    Filipino Restaurant  
2           Golf Course   Fast Food Restaurant  
3     Food & Drink Shop             Food Court  
4  Fast Food Restaurant    Filipino Restaurant

## Model Building

### K Means

Let's cluster the city of london to roughly 5 to make it easier to analyze.

We use the K Means clustering technique to do so.

In [ ]:
# set number of clusters
k_num_clusters = 5

London_grouped_clustering = London_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans_london = KMeans(n_clusters=k_num_clusters, random_state=0).fit(London_grouped_clustering)
kmeans_london

KMeans(n_clusters=5, random_state=0)

### Labelling Clustered Data

In [ ]:
kmeans_london.labels_

array([1, 0, 2, 2, 2, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 3, 1, 1, 1, 1, 1,
       1, 1, 1, 4, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1], dtype=int32)

So our model has labeled the city

In [ ]:
neighborhoods_venues_sorted_london.insert(0, 'Cluster Labels', kmeans_london.labels_ +1)

Join London_merged with our neighbourhood venues sorted to add latitude & longitude for each of the neighborhood to prepare it for plotting

In [ ]:
london_data = london_merged

london_data = london_data.join(neighborhoods_venues_sorted_london.set_index('Neighbourhood'), on='borough')

london_data.head()

borough    town post_code  latitude  longitude  \
0              Bexley, Greenwich   LONDON       SE2  51.49245    0.12127   
1  Ealing, Hammersmith and Fulham  LONDON    W3, W4  51.51324   -0.26746   
6                            City  LONDON       EC3  51.51200   -0.08058   
7                     Westminster  LONDON       WC2  51.51651   -0.11968   
9                         Bromley  LONDON      SE20  51.41009   -0.05683   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               3           Supermarket         Train Station   
1               4         Grocery Store         Train Station   
6               2           Coffee Shop                 Hotel   
7               2                 Hotel           Coffee Shop   
9               1           Supermarket     Convenience Store   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0     Convenience Store           Coffee Shop         Historic Site   
1        Breakfast Spot                  Park     Indian Restaurant   
6    Italian Restaurant  Gym / Fitness Center                   Pub   
7                  Café        Sandwich Place                   Pub   
9                 Hotel         Grocery Store  Fast Food Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0     Fish & Chips Shop    Falafel Restaurant        Farmers Market   
1          Dessert Shop               Exhibit          Dance Studio   
6            Restaurant        Sandwich Place        Scenic Lookout   
7    Italian Restaurant               Theater            Restaurant   
9                  Park                   Pub             Gastropub   

  9th Most Common Venue 10th Most Common Venue  
0  Fast Food Restaurant    Filipino Restaurant  
1            Food Stand             Food Court  
6              Wine Bar      French Restaurant  
7          Burger Joint      French Restaurant  
9                Bistro         Sandwich Place

Drop all the NaN values to prevent data skew

In [ ]:
london_data_nonan = london_data.dropna(subset=['Cluster Labels'])

### Visualizing the clustered neighbourhood

Let's plot the clusters

In [ ]:
map_clusters_london = folium.Map(location=[london_lat_coords, london_lng_coords], zoom_start=12)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(london_data_nonan['latitude'], london_data_nonan['longitude'], london_data_nonan['borough'], london_data_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters_london)
        
map_clusters_london

## Examining our Clusters

Cluster 1

In [ ]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 1, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels 1st Most Common Venue  \
9               LONDON               1           Supermarket   
29   BECKENHAM, LONDON               1           Supermarket   
61              LONDON               1              Pharmacy   
69              LONDON               1              Pharmacy   
100             LONDON               1              Pharmacy   
121             LONDON               1           Bus Station   
127             LONDON               1           Supermarket   
137             LONDON               1              Pharmacy   
217             LONDON               1              Pharmacy   
260             LONDON               1              Pharmacy   
269             LONDON               1              Pharmacy   
316             LONDON               1           Supermarket   
319             LONDON               1              Pharmacy   
357             LONDON               1  Fast Food Restaurant   
360             LONDON               1           Supermarket   
377             LONDON               1              Pharmacy   
442             LONDON               1              Pharmacy   
453             LONDON               1                   Pub   
518             LONDON               1              Pharmacy   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
9       Convenience Store                 Hotel         Grocery Store   
29      Convenience Store                 Hotel         Grocery Store   
61         Discount Store        Sandwich Place  Fast Food Restaurant   
69         Discount Store        Sandwich Place  Fast Food Restaurant   
100        Discount Store        Sandwich Place  Fast Food Restaurant   
121  Gym / Fitness Center        Clothing Store           Supermarket   
127     Convenience Store                 Hotel         Grocery Store   
137        Discount Store        Sandwich Place  Fast Food Restaurant   
217        Discount Store        Sandwich Place  Fast Food Restaurant   
260        Discount Store        Sandwich Place  Fast Food Restaurant   
269        Discount Store        Sandwich Place  Fast Food Restaurant   
316     Convenience Store                 Hotel         Grocery Store   
319        Discount Store        Sandwich Place  Fast Food Restaurant   
357    Chinese Restaurant          Liquor Store         Grocery Store   
360     Convenience Store                 Hotel         Grocery Store   
377        Discount Store        Sandwich Place  Fast Food Restaurant   
442        Discount Store        Sandwich Place  Fast Food Restaurant   
453  Gym / Fitness Center              Pharmacy             Pet Store   
518        Discount Store        Sandwich Place  Fast Food Restaurant   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
9    Fast Food Restaurant                  Park                   Pub   
29   Fast Food Restaurant                  Park                   Pub   
61          Grocery Store     Indian Restaurant                   Pub   
69          Grocery Store     Indian Restaurant                   Pub   
100         Grocery Store     Indian Restaurant                   Pub   
121           Zoo Exhibit     Fish & Chips Shop    Falafel Restaurant   
127  Fast Food Restaurant                  Park                   Pub   
137         Grocery Store     Indian Restaurant                   Pub   
217         Grocery Store     Indian Restaurant                   Pub   
260         Grocery Store     Indian Restaurant                   Pub   
269         Grocery Store     Indian Restaurant                   Pub   
316  Fast Food Restaurant                  Park                   Pub   
319         Grocery Store     Indian Restaurant                   Pub   
357     Convenience Store       Warehouse Store        Discount Store   
360  Fast Food Restaurant                  Park                   Pub   
377         Grocery Store     Indian Restaurant                   Pub   
442         Grocery Store     India

Cluster 2

In [ ]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 2, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels 1st Most Common Venue  \
6                    LONDON               2           Coffee Shop   
7                    LONDON               2                 Hotel   
10                   LONDON               2           Coffee Shop   
12                   LONDON               2           Coffee Shop   
14           BARNET, LONDON               2           Coffee Shop   
..                      ...             ...                   ...   
522                  LONDON               2                  Café   
523  LONDON, WOODFORD GREEN               2                 Hotel   
526                  LONDON               2           Coffee Shop   
527                  LONDON               2                   Pub   
529                  LONDON               2                   Pub   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
6                   Hotel    Italian Restaurant  Gym / Fitness Center   
7             Coffee Shop                  Café        Sandwich Place   
10                    Pub                  Café            Food Truck   
12                    Pub                  Café            Food Truck   
14                   Café         Grocery Store    Italian Restaurant   
..                    ...                   ...                   ...   
522           Coffee Shop         Grocery Store                   Pub   
523                  Café               Theater                   Pub   
526                  Café         Grocery Store    Italian Restaurant   
527         Grocery Store              Bus Stop           Coffee Shop   
529                  Café           Coffee Shop         Grocery Store   

     5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
6                      Pub            Restaurant        Sandwich Place   
7                      Pub    Italian Restaurant               Theater   
10   Vietnamese Restaurant  Gym / Fitness Center    Italian Restaurant   
12   Vietnamese Restaurant  Gym / Fitness Center    Italian Restaurant   
14                     Pub              Bus Stop           Supermarket   
..                     ...                   ...                   ...   
522     Seafood Restaurant              Pharmacy                   Bar   
523                  Plaza                Garden   Monument / Landmark   
526                    Pub              Bus Stop           Supermarket   
527      Indian Restaurant         Historic Site                  Park   
529                  Hotel             Gastropub           Pizza Place   

    8th Most Common Venue 9th Most Common Venue      10th Most Common Venue  
6          Scenic Lookout              Wine Bar           French Restaurant  
7              Restaurant          Burger Joint           French Restaurant  
10                   Park          Cocktail Bar              Breakfast Spot  
12                   Park          Cocktail Bar              Breakfast Spot  
14       Sushi Restaurant              Pharmacy          Turkish Restaurant  
..                    ...                   ...                         ...  
522                Bakery         Metro Station                        Park  
523            Restaurant           Art Gallery                  Steakhouse  
526      Sushi Restaurant              Pharmacy          Turkish Restaurant  
527                  Café        Clothing Store  Construction & Landscaping  
529       Thai Restaurant    Italian Restaurant                        Park  

[281 rows x 12 columns]

Cluster 3

In [ ]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 3, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels 1st Most Common Venue  \
0                 LONDON               3           Supermarket   
45   BEXLEYHEATH, LONDON               3           Supermarket   
124               LONDON               3           Supermarket   
167      LONDON, WELLING               3           Sports Club   
291       LONDON, SIDCUP               3           Supermarket   
458        LONDON, ERITH               3           Sports Club   
506               LONDON               3           Supermarket   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0           Train Station     Convenience Store           Coffee Shop   
45          Historic Site           Coffee Shop     Convenience Store   
124         Historic Site           Coffee Shop     Convenience Store   
167     Convenience Store                  Park         Historic Site   
291         Historic Site           Coffee Shop     Convenience Store   
458     Convenience Store                  Park         Historic Site   
506         Historic Site           Coffee Shop     Convenience Store   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                 Historic Site     Fish & Chips Shop    Falafel Restaurant   
45                Train Station                  Park              Bus Stop   
124               Train Station                  Park              Bus Stop   
167  Construction & Landscaping           Golf Course              Bus Stop   
291               Train Station                  Park              Bus Stop   
458  Construction & Landscaping           Golf Course              Bus Stop   
506               Train Station                  Park              Bus Stop   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                Farmers Market  Fast Food Restaurant    Filipino Restaurant  
45   Construction & Landscaping           Golf Course   Fast Food Restaurant  
124  Construction & Landscaping           Golf Course   Fast Food Restaurant  
167                 Flower Shop     Food & Drink Shop             Food Court  
291  Construction & Landscaping           Golf Course   Fast Food Restaurant  
458                 Flower Shop     Food & Drink Shop             Food Court  
506  Construction & Landscaping           Golf Course   Fast Food Restaurant

Cluster 4

In [ ]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 4, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
1  LONDON               4         Grocery Store         Train Station   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
1        Breakfast Spot                  Park     Indian Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
1          Dessert Shop               Exhibit          Dance Studio   

  9th Most Common Venue 10th Most Common Venue  
1            Food Stand             Food Court

Cluster 5

In [ ]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 5, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels 1st Most Common Venue  \
378  HARROW, STANMOREEDGWARE, LONDON               5                   Gym   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
378         Metro Station                Bakery            Food Stand   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
378            Food Court     Food & Drink Shop           Flower Shop   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
378           Flea Market         Fishing Store            Event Space

# Exploring Paris

### Neighbourhoods of Paris

### Data Collection

We read the json data with pandas.

In [ ]:
!wget -q -O 'france-data.json' https://www.data.gouv.fr/fr/datasets/r/e88c6fda-1d09-42a0-a069-606d3259114e
print("Data Downloaded!")

Data Downloaded!


In [ ]:
paris_raw = pd.read_json('france-data.json')
paris_raw.head()

datasetid  \
0  correspondances-code-insee-code-postal   
1  correspondances-code-insee-code-postal   
2  correspondances-code-insee-code-postal   
3  correspondances-code-insee-code-postal   
4  correspondances-code-insee-code-postal   

                                   recordid  \
0  2bf36b38314b6c39dfbcd09225f97fa532b1fc45   
1  7ee82e74e059b443df18bb79fc5a19b1f05e5a88   
2  e2cd3186f07286705ed482a10b6aebd9de633c81   
3  868bf03527a1d0a9defe5cf4e6fa0a730d725699   
4  21e809b1d4480333c8b6fe7addd8f3b06f343e2c   

                                              fields  \
0  {'code_comm': '645', 'nom_dept': 'ESSONNE', 's...   
1  {'code_comm': '133', 'nom_dept': 'SEINE-ET-MAR...   
2  {'code_comm': '378', 'nom_dept': 'ESSONNE', 's...   
3  {'code_comm': '243', 'nom_dept': 'SEINE-ET-MAR...   
4  {'code_comm': '003', 'nom_dept': 'VAL-DE-MARNE...   

                                            geometry  \
0  {'type': 'Point', 'coordinates': [2.2517129721...   
1  {'type': 'Point', 'coordinates': [3.0529405055...   
2  {'type': 'Point', 'coordinates': [2.1971816504...   
3  {'type': 'Point', 'coordinates': [2.7097808131...   
4  {'type': 'Point', 'coordinates': [2.3335102498...   

                record_timestamp  
0  2016-09-21T00:29:06.175+02:00  
1  2016-09-21T00:29:06.175+02:00  
2  2016-09-21T00:29:06.175+02:00  
3  2016-09-21T00:29:06.175+02:00  
4  2016-09-21T00:29:06.175+02:00

### Data Preprocessing

We break down each of the nested fields and create the dataframe that we need

In [ ]:
paris_field_data = pd.DataFrame()
for f in paris_raw.fields:
    dict_new = f
    paris_field_data = paris_field_data.append(dict_new, ignore_index=True)

paris_field_data.head()

code_arr code_cant code_comm code_dept code_reg  \
0        3        03       645        91       11   
1        3        20       133        77       11   
2        1        09       378        91       11   
3        5        14       243        77       11   
4        3        34       003        94       11   

                              geo_point_2d  \
0  [48.750443119964764, 2.251712972144151]   
1   [48.41256065214989, 3.052940505560729]   
2    [48.52726809075556, 2.19718165044305]   
3  [48.87307018579678, 2.7097808131278462]   
4   [48.80588035965699, 2.333510249842654]   

                                           geo_shape id_geofla insee_com  \
0  {'type': 'Polygon', 'coordinates': [[[2.238024...     16275     91645   
1  {'type': 'Polygon', 'coordinates': [[[3.076046...     31428     77133   
2  {'type': 'Polygon', 'coordinates': [[[2.203466...     30975     91378   
3  {'type': 'Polygon', 'coordinates': [[[2.727542...     17000     77243   
4  {'type': 'Polygon', 'coordinates': [[[2.343851...     32123     94003   

               nom_comm        nom_dept     nom_region  population  \
0  VERRIERES-LE-BUISSON         ESSONNE  ILE-DE-FRANCE        15.5   
1  COURCELLES-EN-BASSEE  SEINE-ET-MARNE  ILE-DE-FRANCE         0.2   
2             MAUCHAMPS         ESSONNE  ILE-DE-FRANCE         0.3   
3       LAGNY-SUR-MARNE  SEINE-ET-MARNE  ILE-DE-FRANCE        20.2   
4               ARCUEIL    VAL-DE-MARNE  ILE-DE-FRANCE        19.5   

  postal_code            statut  superficie  z_moyen  
0       91370    Commune simple       999.0    121.0  
1       77126    Commune simple      1082.0     88.0  
2       91730    Commune simple       313.0    150.0  
3       77400  Chef-lieu canton       579.0     71.0  
4       94110  Chef-lieu canton       232.0     70.0

### Feature Selection
We take the columns that we require, in case of paris it would be the geo_point_2d, nom_dept, nom_comm and postal_code

In [ ]:
df_2 = paris_field_data[['postal_code','nom_comm','nom_dept','geo_point_2d']]
df_2

postal_code              nom_comm           nom_dept  \
0          91370  VERRIERES-LE-BUISSON            ESSONNE   
1          77126  COURCELLES-EN-BASSEE     SEINE-ET-MARNE   
2          91730             MAUCHAMPS            ESSONNE   
3          77400       LAGNY-SUR-MARNE     SEINE-ET-MARNE   
4          94110               ARCUEIL       VAL-DE-MARNE   
...          ...                   ...                ...   
1295       77520     CESSOY-EN-MONTOIS     SEINE-ET-MARNE   
1296       93420            VILLEPINTE  SEINE-SAINT-DENIS   
1297       77130         CANNES-ECLUSE     SEINE-ET-MARNE   
1298       78930              VILLETTE           YVELINES   
1299       95270  LE PLESSIS-LUZARCHES         VAL-D'OISE   

                                 geo_point_2d  
0     [48.750443119964764, 2.251712972144151]  
1      [48.41256065214989, 3.052940505560729]  
2       [48.52726809075556, 2.19718165044305]  
3     [48.87307018579678, 2.7097808131278462]  
4      [48.80588035965699, 2.333510249842654]  
...                                       ...  
1295   [48.50730730461658, 3.138844194183689]  
1296   [48.95902025378707, 2.536306342059409]  
1297   [48.36403767307805, 2.990786679832767]  
1298  [48.92627887061508, 1.6937417245662671]  
1299  [49.09572967201378, 2.4547564431234923]  

[1300 rows x 4 columns]

We have managed to collect 1300 records.

### Feature Engineering

We localize to only include Paris

In [ ]:
df_paris = df_2[df_2['nom_dept'].str.contains('PARIS')].reset_index(drop=True)
df_paris

postal_code                  nom_comm nom_dept  \
0        75009   PARIS-9E-ARRONDISSEMENT    PARIS   
1        75002   PARIS-2E-ARRONDISSEMENT    PARIS   
2        75011  PARIS-11E-ARRONDISSEMENT    PARIS   
3        75003   PARIS-3E-ARRONDISSEMENT    PARIS   
4        75006   PARIS-6E-ARRONDISSEMENT    PARIS   
5        75019  PARIS-19E-ARRONDISSEMENT    PARIS   
6        75020  PARIS-20E-ARRONDISSEMENT    PARIS   
7        75015  PARIS-15E-ARRONDISSEMENT    PARIS   
8        75005   PARIS-5E-ARRONDISSEMENT    PARIS   
9        75007   PARIS-7E-ARRONDISSEMENT    PARIS   
10       75008   PARIS-8E-ARRONDISSEMENT    PARIS   
11       75013  PARIS-13E-ARRONDISSEMENT    PARIS   
12       75012  PARIS-12E-ARRONDISSEMENT    PARIS   
13       75018  PARIS-18E-ARRONDISSEMENT    PARIS   
14       75001  PARIS-1ER-ARRONDISSEMENT    PARIS   
15       75010  PARIS-10E-ARRONDISSEMENT    PARIS   
16       75016  PARIS-16E-ARRONDISSEMENT    PARIS   
17       75004   PARIS-4E-ARRONDISSEMENT    PARIS   
18       75017  PARIS-17E-ARRONDISSEMENT    PARIS   
19       75014  PARIS-14E-ARRONDISSEMENT    PARIS   

                               geo_point_2d  
0    [48.87689616237872, 2.337460241388529]  
1    [48.86790337886785, 2.344107166658533]  
2    [48.85941549762748, 2.378741060237548]  
3    [48.86305413181178, 2.359361058970589]  
4    [48.84896809191946, 2.332670898588416]  
5    [48.88686862295828, 2.384694327870042]  
6    [48.86318677744551, 2.400819826729021]  
7    [48.84015541860987, 2.293559372435076]  
8   [48.844508659617546, 2.349859385560182]  
9    [48.85608259819694, 2.312438687733857]  
10   [48.87252726662346, 2.312582560420059]  
11   [48.82871768452136, 2.362468228516128]  
12   [48.83515623066034, 2.419807034965275]  
13  [48.892735074561706, 2.348711933867703]  
14    [48.8626304851685, 2.336293446550539]  
15   [48.87602855694339, 2.361112904561707]  
16   [48.86039876035177, 2.262099559395783]  
17  [48.854228281954754, 2.357361938142205]  
18   [48.88733716648682, 2.307485559493426]  
19   [48.82899321160942, 2.327100883257538]

In [ ]:
df_paris.shape

(20, 4)

We have managed to bring down the records to just 20 from 1300. Paris data is quite small than initally thought

In [ ]:
df_paris.dtypes

postal_code     object
nom_comm        object
nom_dept        object
geo_point_2d    object
dtype: object

## Gelocations of the Neighbourhoods of Paris

We don't need to get the geo coordinates using an external data source or collect it with the arcgis API call since we already have it stored in the geo_point_2d column as a tuple in the df_paris dataframe.

Checking one of the geo coordinates.

In [ ]:
df_paris['geo_point_2d'][0]

[48.87689616237872, 2.337460241388529]

In [ ]:
temp1 = df_paris['geo_point_2d']
temp1

0      [48.87689616237872, 2.337460241388529]
1      [48.86790337886785, 2.344107166658533]
2      [48.85941549762748, 2.378741060237548]
3      [48.86305413181178, 2.359361058970589]
4      [48.84896809191946, 2.332670898588416]
5      [48.88686862295828, 2.384694327870042]
6      [48.86318677744551, 2.400819826729021]
7      [48.84015541860987, 2.293559372435076]
8     [48.844508659617546, 2.349859385560182]
9      [48.85608259819694, 2.312438687733857]
10     [48.87252726662346, 2.312582560420059]
11     [48.82871768452136, 2.362468228516128]
12     [48.83515623066034, 2.419807034965275]
13    [48.892735074561706, 2.348711933867703]
14      [48.8626304851685, 2.336293446550539]
15     [48.87602855694339, 2.361112904561707]
16     [48.86039876035177, 2.262099559395783]
17    [48.854228281954754, 2.357361938142205]
18     [48.88733716648682, 2.307485559493426]
19     [48.82899321160942, 2.327100883257538]
Name: geo_point_2d, dtype: object

In [ ]:
paris_latlng = df_paris['geo_point_2d'].astype('str')

Spliting the geo_point_2d column into latitude and longitude.

### Latitude

In [ ]:
paris_lat = paris_latlng.apply(lambda x: x.split(',')[0])
paris_lat = paris_lat.apply(lambda x: x.lstrip('['))
paris_lat

0      48.87689616237872
1      48.86790337886785
2      48.85941549762748
3      48.86305413181178
4      48.84896809191946
5      48.88686862295828
6      48.86318677744551
7      48.84015541860987
8     48.844508659617546
9      48.85608259819694
10     48.87252726662346
11     48.82871768452136
12     48.83515623066034
13    48.892735074561706
14      48.8626304851685
15     48.87602855694339
16     48.86039876035177
17    48.854228281954754
18     48.88733716648682
19     48.82899321160942
Name: geo_point_2d, dtype: object

### Longitude

In [ ]:
paris_lng = paris_latlng.apply(lambda x: x.split(',')[1])
paris_lng = paris_lng.apply(lambda x: x.rstrip(']'))
paris_lng

0      2.337460241388529
1      2.344107166658533
2      2.378741060237548
3      2.359361058970589
4      2.332670898588416
5      2.384694327870042
6      2.400819826729021
7      2.293559372435076
8      2.349859385560182
9      2.312438687733857
10     2.312582560420059
11     2.362468228516128
12     2.419807034965275
13     2.348711933867703
14     2.336293446550539
15     2.361112904561707
16     2.262099559395783
17     2.357361938142205
18     2.307485559493426
19     2.327100883257538
Name: geo_point_2d, dtype: object

In [ ]:
paris_geo_lat  = pd.DataFrame(paris_lat.astype(float))
paris_geo_lat.columns=['Latitude']
paris_geo_lng = pd.DataFrame(paris_lng.astype(float))
paris_geo_lng.columns=['Longitude']

Preparing our combined data by dropping the geo_point_2d column from our previously stored df_paris and concatenating with the latitude and longitude extracted from it

In [ ]:
paris_combined_data = pd.concat([df_paris.drop('geo_point_2d', axis=1), paris_geo_lat, paris_geo_lng], axis=1)
paris_combined_data

postal_code                  nom_comm nom_dept   Latitude  Longitude
0        75009   PARIS-9E-ARRONDISSEMENT    PARIS  48.876896   2.337460
1        75002   PARIS-2E-ARRONDISSEMENT    PARIS  48.867903   2.344107
2        75011  PARIS-11E-ARRONDISSEMENT    PARIS  48.859415   2.378741
3        75003   PARIS-3E-ARRONDISSEMENT    PARIS  48.863054   2.359361
4        75006   PARIS-6E-ARRONDISSEMENT    PARIS  48.848968   2.332671
5        75019  PARIS-19E-ARRONDISSEMENT    PARIS  48.886869   2.384694
6        75020  PARIS-20E-ARRONDISSEMENT    PARIS  48.863187   2.400820
7        75015  PARIS-15E-ARRONDISSEMENT    PARIS  48.840155   2.293559
8        75005   PARIS-5E-ARRONDISSEMENT    PARIS  48.844509   2.349859
9        75007   PARIS-7E-ARRONDISSEMENT    PARIS  48.856083   2.312439
10       75008   PARIS-8E-ARRONDISSEMENT    PARIS  48.872527   2.312583
11       75013  PARIS-13E-ARRONDISSEMENT    PARIS  48.828718   2.362468
12       75012  PARIS-12E-ARRONDISSEMENT    PARIS  48.835156   2.419807
13       75018  PARIS-18E-ARRONDISSEMENT    PARIS  48.892735   2.348712
14       75001  PARIS-1ER-ARRONDISSEMENT    PARIS  48.862630   2.336293
15       75010  PARIS-10E-ARRONDISSEMENT    PARIS  48.876029   2.361113
16       75016  PARIS-16E-ARRONDISSEMENT    PARIS  48.860399   2.262100
17       75004   PARIS-4E-ARRONDISSEMENT    PARIS  48.854228   2.357362
18       75017  PARIS-17E-ARRONDISSEMENT    PARIS  48.887337   2.307486
19       75014  PARIS-14E-ARRONDISSEMENT    PARIS  48.828993   2.327101

### Co-ordinates for Paris

In [ ]:
paris = geocode(address='Paris, France, FR')[0]
paris_lng_coords = paris['location']['x']
paris_lat_coords = paris['location']['y']
print("The geolocation of Paris: ", paris_lat_coords, paris_lng_coords)

The geolocation of Paris:  48.85341000000005 2.3488000000000397


## Visualize the Map of Paris

In [ ]:
# Creating the map of Paris
map_Paris= folium.Map(location=[paris_lat_coords, paris_lng_coords], zoom_start=12)
map_Paris

# adding markers to map
for latitude, longitude, borough, town in zip(paris_combined_data['Latitude'], paris_combined_data['Longitude'], paris_combined_data['nom_comm'], paris_combined_data['nom_dept']):
    label = '{}, {}'.format(town, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='Blue',
        fill=True,
        fill_opacity=0.8
        ).add_to(map_Paris)  
    
map_Paris

### Venues in Paris

Using our previously defined function. Let's get the neaby venues present in each neighbourhood of Paris

In [ ]:
venues_in_Paris = getNearbyVenues(paris_combined_data['nom_comm'], paris_combined_data['Latitude'], paris_combined_data['Longitude'])

PARIS-9E-ARRONDISSEMENT
PARIS-2E-ARRONDISSEMENT
PARIS-11E-ARRONDISSEMENT
PARIS-3E-ARRONDISSEMENT
PARIS-6E-ARRONDISSEMENT
PARIS-19E-ARRONDISSEMENT
PARIS-20E-ARRONDISSEMENT
PARIS-15E-ARRONDISSEMENT
PARIS-5E-ARRONDISSEMENT
PARIS-7E-ARRONDISSEMENT
PARIS-8E-ARRONDISSEMENT
PARIS-13E-ARRONDISSEMENT
PARIS-12E-ARRONDISSEMENT
PARIS-18E-ARRONDISSEMENT
PARIS-1ER-ARRONDISSEMENT
PARIS-10E-ARRONDISSEMENT
PARIS-16E-ARRONDISSEMENT
PARIS-4E-ARRONDISSEMENT
PARIS-17E-ARRONDISSEMENT
PARIS-14E-ARRONDISSEMENT


Sampling our Data

In [ ]:
venues_in_Paris.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0  PARIS-9E-ARRONDISSEMENT               48.876896                  2.33746   
1  PARIS-9E-ARRONDISSEMENT               48.876896                  2.33746   
2  PARIS-9E-ARRONDISSEMENT               48.876896                  2.33746   
3  PARIS-9E-ARRONDISSEMENT               48.876896                  2.33746   
4  PARIS-9E-ARRONDISSEMENT               48.876896                  2.33746   

                    Venue         Venue Category  
0     Place Saint-Georges                  Plaza  
1                     Qee   Gym / Fitness Center  
2  Le Bouclier de Bacchus               Wine Bar  
3               Aji Dulce  Venezuelan Restaurant  
4                     RAP           Gourmet Shop

In [ ]:
venues_in_Paris.shape

(1306, 5)

We have managed to collect 1306 venue records for the neighbourhoods in Paris

### Grouping by Venue Categories

We need to now see how many Venue Categories are there for further processing

In [ ]:
venues_in_Paris.groupby('Venue Category').max()

Neighbourhood  Neighbourhood Latitude  \
Venue Category                                                             
Afghan Restaurant       PARIS-11E-ARRONDISSEMENT               48.859415   
African Restaurant       PARIS-9E-ARRONDISSEMENT               48.876896   
American Restaurant     PARIS-19E-ARRONDISSEMENT               48.892735   
Antique Shop             PARIS-9E-ARRONDISSEMENT               48.876896   
Argentinian Restaurant   PARIS-3E-ARRONDISSEMENT               48.863054   
...                                          ...                     ...   
Wine Bar                 PARIS-9E-ARRONDISSEMENT               48.892735   
Wine Shop                PARIS-3E-ARRONDISSEMENT               48.876029   
Women's Store            PARIS-6E-ARRONDISSEMENT               48.867903   
Zoo                     PARIS-12E-ARRONDISSEMENT               48.835156   
Zoo Exhibit             PARIS-12E-ARRONDISSEMENT               48.835156   

                        Neighbourhood Longitude  \
Venue Category                                    
Afghan Restaurant                      2.378741   
African Restaurant                     2.361113   
American Restaurant                    2.384694   
Antique Shop                           2.337460   
Argentinian Restaurant                 2.359361   
...                                         ...   
Wine Bar                               2.400820   
Wine Shop                              2.361113   
Women's Store                          2.344107   
Zoo                                    2.419807   
Zoo Exhibit                            2.419807   

                                                           Venue  
Venue Category                                                    
Afghan Restaurant                                    Afghanistan  
African Restaurant                             Wally Le Saharien  
American Restaurant                                     Harper's  
Antique Shop                             Hôtel des Ventes Drouot  
Argentinian Restaurant                                     Anahi  
...                                                          ...  
Wine Bar                                        Vingt Vins d'Art  
Wine Shop                                         Trois Fois Vin  
Women's Store                               L'Appartement Sézane  
Zoo                                     Parc zoologique de Paris  
Zoo Exhibit             Grande Serre du Parc Zoologique de Paris  

[204 rows x 4 columns]

204 records in Paris. Paris is also multicultural and diverse as London

### One Hot Encoding

We need to Encode our venue categories to get a better result for our clustering

In [ ]:
Paris_venue_cat = pd.get_dummies(venues_in_Paris[['Venue Category']], prefix="", prefix_sep="")
Paris_venue_cat

Afghan Restaurant  African Restaurant  American Restaurant  \
0                     0                   0                    0   
1                     0                   0                    0   
2                     0                   0                    0   
3                     0                   0                    0   
4                     0                   0                    0   
...                 ...                 ...                  ...   
1301                  0                   0                    0   
1302                  0                   0                    0   
1303                  0                   0                    0   
1304                  0                   0                    0   
1305                  0                   0                    0   

      Antique Shop  Argentinian Restaurant  Art Gallery  Art Museum  \
0                0                       0            0           0   
1                0                       0            0           0   
2                0                       0            0           0   
3                0                       0            0           0   
4                0                       0            0           0   
...            ...                     ...          ...         ...   
1301             0                       0            0           0   
1302             0                       0            0           0   
1303             0                       0            0           0   
1304             0                       0            0           0   
1305             0                       0            0           0   

      Arts & Crafts Store  Asian Restaurant  Athletics & Sports  ...  \
0                       0                 0                   0  ...   
1                       0                 0                   0  ...   
2                       0                 0                   0  ...   
3                       0                 0                   0  ...   
4                       0                 0                   0  ...   
...                   ...               ...                 ...  ...   
1301                    0                 0                   0  ...   
1302                    0                 0                   0  ...   
1303                    0                 0                   0  ...   
1304                    0                 0                   0  ...   
1305                    0                 0                   0  ...   

      Turkish Restaurant  Udon Restaurant  Vegetarian / Vegan Restaurant  \
0                      0                0                              0   
1                      0                0                              0   
2                      0                0                              0   
3                      0                0                              0   
4                      0                0                              0   
...                  ...              ...                            ...   
1301                   0                0                              0   
1302                   0                0                              0   
1303                   0                0                              0   
1304                   0                0                              0   
1305                   0                0                              0   

      Venezuelan Restaurant  Vietnamese Restaurant  Wine Bar  Wine Shop  \
0                         0                      0         0          0   
1                         0                      0         0          0   
2                         0                      0         1          0   
3                         1                      0         0          0   
4                         0                      0         0          0   
...                     ...                    ...       ...        ...   
1301                      0      

Adding Neighbourhoods to our Data

In [ ]:
Paris_venue_cat['Neighbourhood'] = venues_in_Paris['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [Paris_venue_cat.columns[-1]] + list(Paris_venue_cat.columns[:-1])
Paris_venue_cat = Paris_venue_cat[fixed_columns]

Paris_venue_cat.head()

Neighbourhood  Afghan Restaurant  African Restaurant  \
0  PARIS-9E-ARRONDISSEMENT                  0                   0   
1  PARIS-9E-ARRONDISSEMENT                  0                   0   
2  PARIS-9E-ARRONDISSEMENT                  0                   0   
3  PARIS-9E-ARRONDISSEMENT                  0                   0   
4  PARIS-9E-ARRONDISSEMENT                  0                   0   

   American Restaurant  Antique Shop  Argentinian Restaurant  Art Gallery  \
0                    0             0                       0            0   
1                    0             0                       0            0   
2                    0             0                       0            0   
3                    0             0                       0            0   
4                    0             0                       0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  ...  Turkish Restaurant  \
0           0                    0                 0  ...                   0   
1           0                    0                 0  ...                   0   
2           0                    0                 0  ...                   0   
3           0                    0                 0  ...                   0   
4           0                    0                 0  ...                   0   

   Udon Restaurant  Vegetarian / Vegan Restaurant  Venezuelan Restaurant  \
0                0                              0                      0   
1                0                              0                      0   
2                0                              0                      0   
3                0                              0                      1   
4                0                              0                      0   

   Vietnamese Restaurant  Wine Bar  Wine Shop  Women's Store  Zoo  Zoo Exhibit  
0                      0         0          0              0    0            0  
1                      0         0          0              0    0            0  
2                      0         1          0              0    0            0  
3                      0         0          0              0    0            0  
4                      0         0          0              0    0            0  

[5 rows x 205 columns]

### Venue categories mean value

We will group the Neighbourhoods and calculate the mean venue categories value in each Neighbourhood

In [ ]:
Paris_grouped = Paris_venue_cat.groupby('Neighbourhood').mean().reset_index()
Paris_grouped.head()

Neighbourhood  Afghan Restaurant  African Restaurant  \
0  PARIS-10E-ARRONDISSEMENT            0.00000                0.02   
1  PARIS-11E-ARRONDISSEMENT            0.02381                0.00   
2  PARIS-12E-ARRONDISSEMENT            0.00000                0.00   
3  PARIS-13E-ARRONDISSEMENT            0.00000                0.00   
4  PARIS-14E-ARRONDISSEMENT            0.00000                0.00   

   American Restaurant  Antique Shop  Argentinian Restaurant  Art Gallery  \
0                  0.0           0.0                     0.0          0.0   
1                  0.0           0.0                     0.0          0.0   
2                  0.0           0.0                     0.0          0.0   
3                  0.0           0.0                     0.0          0.0   
4                  0.0           0.0                     0.0          0.0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  ...  Turkish Restaurant  \
0     0.00000                  0.0          0.030000  ...                 0.0   
1     0.02381                  0.0          0.023810  ...                 0.0   
2     0.00000                  0.0          0.000000  ...                 0.0   
3     0.00000                  0.0          0.189655  ...                 0.0   
4     0.00000                  0.0          0.000000  ...                 0.0   

   Udon Restaurant  Vegetarian / Vegan Restaurant  Venezuelan Restaurant  \
0              0.0                       0.010000                    0.0   
1              0.0                       0.047619                    0.0   
2              0.0                       0.000000                    0.0   
3              0.0                       0.000000                    0.0   
4              0.0                       0.000000                    0.0   

   Vietnamese Restaurant  Wine Bar  Wine Shop  Women's Store  Zoo  Zoo Exhibit  
0               0.000000   0.02000       0.02            0.0  0.0          0.0  
1               0.023810   0.02381       0.00            0.0  0.0          0.0  
2               0.000000   0.00000       0.00            0.0  0.2          0.2  
3               0.206897   0.00000       0.00            0.0  0.0          0.0  
4               0.000000   0.00000       0.00            0.0  0.0          0.0  

[5 rows x 205 columns]

### Top venue categories
Reusing our previously defined function to get the top venue categories in the neighbourhoods of Paris.

In [ ]:
# create a new dataframe for Paris
neighborhoods_venues_sorted_paris = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted_paris['Neighbourhood'] = Paris_grouped['Neighbourhood']

for ind in np.arange(Paris_grouped.shape[0]):
    neighborhoods_venues_sorted_paris.iloc[ind, 1:] = return_most_common_venues(Paris_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted_paris.head()

Neighbourhood  1st Most Common Venue 2nd Most Common Venue  \
0  PARIS-10E-ARRONDISSEMENT      French Restaurant                 Hotel   
1  PARIS-11E-ARRONDISSEMENT             Restaurant     French Restaurant   
2  PARIS-12E-ARRONDISSEMENT            Zoo Exhibit           Supermarket   
3  PARIS-13E-ARRONDISSEMENT  Vietnamese Restaurant      Asian Restaurant   
4  PARIS-14E-ARRONDISSEMENT      French Restaurant                 Hotel   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0           Coffee Shop                  Café                Bistro   
1                  Café           Pastry Shop          Cocktail Bar   
2   Monument / Landmark                Bistro                   Zoo   
3       Thai Restaurant    Chinese Restaurant     French Restaurant   
4   Japanese Restaurant     Food & Drink Shop                  Café   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0             Pizza Place     Indian Restaurant      Asian Restaurant   
1                  Bistro    Italian Restaurant                Bakery   
2  Argentinian Restaurant                Garden    Frozen Yogurt Shop   
3               Juice Bar                 Hotel                  Park   
4                  Bistro                Bakery  Fast Food Restaurant   

           9th Most Common Venue 10th Most Common Venue  
0             Italian Restaurant    Japanese Restaurant  
1  Vegetarian / Vegan Restaurant    Moroccan Restaurant  
2              French Restaurant               Fountain  
3                       Bus Stop   Cambodian Restaurant  
4                    Pizza Place                  Plaza

## Model Building

### K Means

Let's cluster the city of Paris to roughly 5 to make it easier to analyze.

We use the K Means clustering technique to do so.

In [ ]:
# set number of clusters
k_num_clusters = 5

Paris_grouped_clustering = Paris_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans_Paris = KMeans(n_clusters=k_num_clusters, random_state=0).fit(Paris_grouped_clustering)
kmeans_Paris

KMeans(n_clusters=5, random_state=0)

### Labelling Clustered Data

In [ ]:
kmeans_Paris.labels_

array([2, 2, 3, 4, 1, 2, 0, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 2],
      dtype=int32)

So our model has labeled the city, we insert it in our data.

In [ ]:
neighborhoods_venues_sorted_paris.insert(0, 'Cluster Labels', kmeans_Paris.labels_ +1)

Join paris_combined_data with our neighbourhood venues sorted to add latitude & longitude for each of the neighborhood to prepare it for plotting

In [ ]:
paris_data = paris_combined_data

paris_data = paris_data.join(neighborhoods_venues_sorted_paris.set_index('Neighbourhood'), on='nom_comm')

paris_data.head()

postal_code                  nom_comm nom_dept   Latitude  Longitude  \
0       75009   PARIS-9E-ARRONDISSEMENT    PARIS  48.876896   2.337460   
1       75002   PARIS-2E-ARRONDISSEMENT    PARIS  48.867903   2.344107   
2       75011  PARIS-11E-ARRONDISSEMENT    PARIS  48.859415   2.378741   
3       75003   PARIS-3E-ARRONDISSEMENT    PARIS  48.863054   2.359361   
4       75006   PARIS-6E-ARRONDISSEMENT    PARIS  48.848968   2.332671   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               3     French Restaurant                 Hotel   
1               3     French Restaurant          Cocktail Bar   
2               3            Restaurant     French Restaurant   
3               3           Coffee Shop                Bakery   
4               3     French Restaurant        Chocolate Shop   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                Bistro                Bakery   Japanese Restaurant   
1              Wine Bar                Bistro       Thai Restaurant   
2                  Café           Pastry Shop          Cocktail Bar   
3     French Restaurant   Japanese Restaurant          Burger Joint   
4                Bistro                Bakery           Pastry Shop   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0              Wine Bar                Lounge          Cocktail Bar   
1                Bakery    Italian Restaurant                   Bar   
2                Bistro    Italian Restaurant                Bakery   
3                Bistro          Gourmet Shop    Italian Restaurant   
4                 Hotel              Wine Bar            Restaurant   

           9th Most Common Venue 10th Most Common Venue  
0                    Pizza Place            Cheese Shop  
1                    Salad Place                  Hotel  
2  Vegetarian / Vegan Restaurant    Moroccan Restaurant  
3                           Café         Sandwich Place  
4                       Fountain                  Plaza

Drop all the NaN values to prevent data skew

In [ ]:
paris_data_nonan = paris_data.dropna(subset=['Cluster Labels'])

### Visualizing the clustered neighbourhood

Let's plot the clusters

In [ ]:
map_clusters_paris = folium.Map(location=[paris_lat_coords, paris_lng_coords], zoom_start=12)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(paris_data_nonan['Latitude'], paris_data_nonan['Longitude'], paris_data_nonan['nom_comm'], paris_data_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + ' ' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)],
        fill_opacity=0.8
        ).add_to(map_clusters_paris)
        
map_clusters_paris

# Examining our Clusters

Cluster 1

In [ ]:
paris_data_nonan.loc[paris_data_nonan['Cluster Labels'] == 1, paris_data_nonan.columns[[1] + list(range(5, paris_data_nonan.shape[1]))]]

nom_comm  Cluster Labels 1st Most Common Venue  \
16  PARIS-16E-ARRONDISSEMENT               1                 Plaza   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
16                  Lake                  Park           Bus Station   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
16                  Pool     French Restaurant         Boat or Ferry   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
16            Art Museum              Bus Stop     Falafel Restaurant

Cluster 2

In [ ]:
paris_data_nonan.loc[paris_data_nonan['Cluster Labels'] == 2, paris_data_nonan.columns[[1] + list(range(5, paris_data_nonan.shape[1]))]]

nom_comm  Cluster Labels 1st Most Common Venue  \
9    PARIS-7E-ARRONDISSEMENT               2     French Restaurant   
10   PARIS-8E-ARRONDISSEMENT               2     French Restaurant   
18  PARIS-17E-ARRONDISSEMENT               2     French Restaurant   
19  PARIS-14E-ARRONDISSEMENT               2     French Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
9                  Hotel                  Café    Italian Restaurant   
10                 Hotel                   Spa           Art Gallery   
18                 Hotel    Italian Restaurant   Japanese Restaurant   
19                 Hotel   Japanese Restaurant     Food & Drink Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
9                  Plaza        History Museum           Coffee Shop   
10               Theater          Cocktail Bar                Resort   
18                Bakery                  Café            Restaurant   
19                  Café                Bistro                Bakery   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
9             Art Museum          Cocktail Bar    Japanese Restaurant  
10           Coffee Shop                   Bar                   Park  
18                Bistro                 Plaza           Concert Hall  
19  Fast Food Restaurant           Pizza Place                  Plaza

Cluster 3

In [ ]:
paris_data_nonan.loc[paris_data_nonan['Cluster Labels'] == 3, paris_data_nonan.columns[[1] + list(range(5, paris_data_nonan.shape[1]))]]

nom_comm  Cluster Labels 1st Most Common Venue  \
0    PARIS-9E-ARRONDISSEMENT               3     French Restaurant   
1    PARIS-2E-ARRONDISSEMENT               3     French Restaurant   
2   PARIS-11E-ARRONDISSEMENT               3            Restaurant   
3    PARIS-3E-ARRONDISSEMENT               3           Coffee Shop   
4    PARIS-6E-ARRONDISSEMENT               3     French Restaurant   
5   PARIS-19E-ARRONDISSEMENT               3     French Restaurant   
6   PARIS-20E-ARRONDISSEMENT               3                 Plaza   
7   PARIS-15E-ARRONDISSEMENT               3    Italian Restaurant   
8    PARIS-5E-ARRONDISSEMENT               3     French Restaurant   
13  PARIS-18E-ARRONDISSEMENT               3     French Restaurant   
14  PARIS-1ER-ARRONDISSEMENT               3     French Restaurant   
15  PARIS-10E-ARRONDISSEMENT               3     French Restaurant   
17   PARIS-4E-ARRONDISSEMENT               3     French Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Hotel                Bistro                Bakery   
1           Cocktail Bar              Wine Bar                Bistro   
2      French Restaurant                  Café           Pastry Shop   
3                 Bakery     French Restaurant   Japanese Restaurant   
4         Chocolate Shop                Bistro                Bakery   
5                    Bar                 Hotel           Supermarket   
6                 Bistro                Bakery   Japanese Restaurant   
7                  Hotel     French Restaurant   Japanese Restaurant   
8                  Hotel    Italian Restaurant                Bakery   
13                   Bar           Pizza Place                Bistro   
14   Japanese Restaurant                 Hotel                 Plaza   
15                 Hotel           Coffee Shop                  Café   
17        Ice Cream Shop        Clothing Store                 Hotel   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0    Japanese Restaurant              Wine Bar                Lounge   
1        Thai Restaurant                Bakery    Italian Restaurant   
2           Cocktail Bar                Bistro    Italian Restaurant   
3           Burger Joint                Bistro          Gourmet Shop   
4            Pastry Shop                 Hotel              Wine Bar   
5                 Bistro              Beer Bar               Brewery   
6      French Restaurant                   Bar           Pizza Place   
7    Lebanese Restaurant            Restaurant           Coffee Shop   
8                   Café                 Plaza                   Bar   
13                  Café            Restaurant    Italian Restaurant   
14           Coffee Shop      Ramen Restaurant    Italian Restaurant   
15                Bistro           Pizza Place     Indian Restaurant   
17                  Park           Pastry Shop      Pedestrian Plaza   

   8th Most Common Venue          9th Most Common Venue 10th Most Common Venue  
0           Cocktail Bar                    Pizza Place            Cheese Shop  
1                    Bar                    Salad Place                  Hotel  
2                 Bakery  Vegetarian / Vegan Restaurant    Moroccan Restaurant  
3     Italian Restaurant                           Café         Sandwich Place  
4             Restaurant                       Fountain                  Plaza  
5     Seafood Restaurant            Japanese Restaurant         Sandwich Place  
6                   Park                          Hotel     Italian Restaurant  
7        Thai Restaurant                         Bistro                   Park  
8                    Pub                    Coffee Shop         Science Museum  
13                 Plaza                          Hotel  Vietnamese Restaurant  
14                  Café                     Art Museum      Korean Restaurant  
15      Asian Restaurant             Italian Rest

Cluster 4

In [ ]:
paris_data_nonan.loc[paris_data_nonan['Cluster Labels'] == 4, paris_data_nonan.columns[[1] + list(range(5, paris_data_nonan.shape[1]))]]

nom_comm  Cluster Labels 1st Most Common Venue  \
12  PARIS-12E-ARRONDISSEMENT               4           Zoo Exhibit   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
12           Supermarket   Monument / Landmark                Bistro   

   5th Most Common Venue   6th Most Common Venue 7th Most Common Venue  \
12                   Zoo  Argentinian Restaurant                Garden   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
12    Frozen Yogurt Shop     French Restaurant               Fountain

Cluster 5

In [ ]:
paris_data_nonan.loc[paris_data_nonan['Cluster Labels'] == 5, paris_data_nonan.columns[[1] + list(range(5, paris_data_nonan.shape[1]))]]

nom_comm  Cluster Labels  1st Most Common Venue  \
11  PARIS-13E-ARRONDISSEMENT               5  Vietnamese Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
11      Asian Restaurant       Thai Restaurant    Chinese Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
11     French Restaurant             Juice Bar                 Hotel   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
11                  Park              Bus Stop   Cambodian Restaurant

# Results and Discussion

The neighbourhoods of London are very mulitcultural. There are a lot of different cusines including Indian, Italian, Turkish and Chinese. London seems to take a step further in this direction by having a lot of Restaurants, bars, juice bars, coffee shops, Fish and Chips shop and Breakfast spots. It has a lot of shopping options too with that of the Flea markets, flower shops, fish markets, Fishing stores, clothing stores. The main modes of transport seem to be Buses and trains. For leisure, the neighbourhoods are set up to have lots of parks, golf courses, zoo, gyms and Historic sites.

Overall, the city of London offers a multicultural, diverse and certainly an entertaining experience.

Paris is relatively small in size geographically. It has a wide variety of cusines and eateries including French, Thai, Cambodian, Asian, Chinese etc. There are a lot of hangout spots including many Restaurants and Bars. Paris has a lot of Bistro's. Different means of public transport in Paris which includes buses, bikes, boats or ferries. For leisure and sight seeing, there are a lot of Plazas, Trails, Parks, Historic sites, clothing shops, Art galleries and Museums. Overall, Paris seems like the relaxing vacation spot with a mix of lakes, historic spots and a wide variety of cusines to try out.

# Conclusion

The purpose of this project was to explore the cities of London and Paris and see how attractive it is to potential tourists and migrants. We explored both the cities based on their postal codes and then extrapolated the common venues present in each of the neighbourhoods finally concluding with clustering similar neighbourhoods together.

We could see that each of the neighbourhoods in both the cities have a wide variety of experiences to offer which is unique in it's own way. The cultural diversity is quite evident which also gives the feeling of a sense of inclusion.

Both Paris and London seem to offer a vacation stay or a romantic gateaway with a lot of places to explore, beautiful landscapes and a wide variety of culture.Overall, it's upto the stakeholders to decide which experience they would prefer more and which would more to their liking.